taichi/analysis/same_statements.cpp

```diff
@@ -97,22 +120,57 @@ class IRNodeComparator : public IRVisitor {
       same = false;
       return;
     }
+    auto other = other_node->as<Stmt>();
 
     // If two identical statements can have different values, return false.
-    if (check_same_value_ && !stmt->is_container_statement() &&
-        !stmt->common_statement_eliminable()) {
-      same = false;
-      return;
+    // TODO: two identical GlobalPtrStmts cannot have different values,
+    //  but GlobalPtrStmt::common_statement_eliminable() is false.
+    if (check_same_value_ && stmt != other && !stmt->is_container_statement() &&
+        !stmt->common_statement_eliminable() && !stmt->is<GlobalPtrStmt>()) {
+      if (all_states_can_be_modified_) {
+        same = false;
+        return;
+      } else {
+        bool same_value = false;
+        if (auto global_load = stmt->cast<GlobalLoadStmt>()) {
+          if (auto global_ptr = global_load->ptr->cast<GlobalPtrStmt>()) {
+            TI_ASSERT(global_ptr->width() == 1);
+            if (possibly_modified_states_.count(ir_bank_->get_async_state(
+                    global_ptr->snodes[0], AsyncState::Type::value)) == 0) {
+              same_value = true;
+            }
+          }
+          // TODO: other cases?
+        }
+        if (!same_value) {
+          same = false;
+          return;
+        }
+      }
     }
     // Note that we do not need to test !stmt2->common_statement_eliminable()
     // because if this condition does not hold,
     // same_statements(stmt1, stmt2) returns false anyway.
 
     // field check
```

Comment:
nit: move this comment to L170?

Reply:
I think it should be L130 though... `stmt->common_statement_eliminable()` is called there.

Reply:
Oh I see.